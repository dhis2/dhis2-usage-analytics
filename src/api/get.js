if (!process.env.REACT_APP_DHIS2_BASE_URL) {
  throw new Error('The environment variable REACT_APP_DHIS2_BASE_URL must be set');
}

const url = process.env.REACT_APP_DHIS2_BASE_URL;
const endpoint = `${url}/api`;
const defaultConfig = {
  method: 'GET',
  credentials: 'include',
}

export function get(path, config = defaultConfig) {
  if (!path) {
    return Promise.reject(new Error('A path must be passed'));
  }

  return fetch(`${endpoint}/${path}`, config)
    .then(response => {
      if (response.ok) {
        return response;
      }
    
      throw new Error(response.statusText)
    })
}

export function getJSON(path) {
  return get(path)
    .then(response => response.json())
    .then(data => {
      if (data.status === 'ERROR') {
        throw new Error(data.message)
      } else {
        return data
      }
    })
}
