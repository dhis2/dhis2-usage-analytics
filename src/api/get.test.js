import { get, getJSON } from './get'

window.fetch = jest.fn()

describe('get', () => {
  it('rejects with the expected error message if path is not specified', () => {
    return expect(get()).rejects.toMatchSnapshot();
  })

  it('calls fetch with the expected path', () => {
    const actual = 'path';
    const expected = {
      ok: true,
      path: `${process.env.REACT_APP_DHIS2_BASE_URL}/api/${actual}`
    }
    window.fetch.mockImplementationOnce((path) => Promise.resolve({ ok: true, path }))

    return expect(get('path')).resolves.toEqual(expected);
  })

  it('allows the config to be overridden', () => {
    window.fetch.mockImplementationOnce((path, config) => Promise.resolve({ ok: true, config }))
    const actual = 'config'
    const expected = {
      ok: true,
      config: actual
    };

    return expect(get('path', actual)).resolves.toEqual(expected);
  })

  it('rejects with the expected error message if fetch resolves but the response is not ok', () => {
    window.fetch.mockImplementationOnce(() => Promise.resolve({ statusText: 'The response is not ok'}))

    return expect(get('path')).rejects.toMatchSnapshot();
  })
})

describe('getJSON', () => {
  it('returns the data when everything is ok', () => {
    const expected = 'data';
    const response = {
      ok: true,
      json: () => expected
    }
    window.fetch.mockImplementationOnce(() => Promise.resolve(response))

    return expect(getJSON('path')).resolves.toEqual(expected);
  })

  it('rejects with the expected error message if the json is not ok', () => {
    const expected = 'The JSON is not ok'
    const response = {
      ok: true,
      json: () => ({ status: 'ERROR', message: expected })
    }
    window.fetch.mockImplementationOnce(() => Promise.resolve(response))

    return expect(getJSON('path')).rejects.toMatchSnapshot()
  })
})
