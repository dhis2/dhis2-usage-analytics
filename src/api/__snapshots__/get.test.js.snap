// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get rejects with the expected error message if baseUrl is not specified 1`] = `[Error: A baseUrl must be passed]`;

exports[`get rejects with the expected error message if fetch resolves but the response is not ok 1`] = `[Error: The response is not ok]`;

exports[`get rejects with the expected error message if path is not specified 1`] = `[Error: A path must be passed]`;

exports[`getJSON rejects with the expected error message if the json is not ok 1`] = `[Error: The JSON is not ok]`;
